extmod/uasyncio/lock.py

```diff
@@ -0,0 +1,48 @@
+# MicroPython uasyncio module
+# MIT license; Copyright (c) 2019-2020 Damien P. George
+
+from . import core
+
+# Lock class for primitive mutex capability
+class Lock:
+    def __init__(self):
+        self.state = 0  # 0=unlocked; 1=unlocked but waiting task pending resume; 2=locked
+        self.waiting = core.TaskQueue()  # Queue of Tasks waiting to acquire this Lock
+
+    def locked(self):
+        return self.state == 2
+
+    def release(self):
+        if self.state != 2:
+            raise RuntimeError
+        if self.waiting.peek():
+            # Task(s) waiting on lock, schedule first Task
+            core._task_queue.push_head(self.waiting.pop_head())
+            self.state = 1
+        else:
+            # No Task waiting so unlock
+            self.state = 0
+
+    async def acquire(self):
+        if self.state != 0 or self.waiting.peek():
```

Comment:
Why the `or self.waiting.peek()`? The only place where `state` is set to 0 is 3 lines higher and it has a `if not self.waiting.peek()` predicate.

Reply:
I assume this was an attempt to fix a bug I found in the Lock implementation.
In my version of Lock this line looks like this:
`if self._locked or self._awt:`
where self._awt stores the task that is going to acquire the lock next.

The bug is that if you have 5 tasks waiting for a lock and cancel 2 of them, then 2 tasks get scheduled and one of them acquires a locked lock.
This bug is still present in this code and therefore this implementation will need to see some changes anyway. Was just waiting for this to be merged to create an issue/PR.

Reply:
Here's a test case for the bug:
```
import sys
if sys.platform == 'linux': sys.path.append('/home/src/esp32/micropython/extmod')
import uasyncio as asyncio
lock = asyncio.Lock()
async def locker(i):
    try:
        print("acquiring", i)
        async with lock:
            print("got", i)
            await asyncio.sleep_ms(100)
            print("releasing", i)
    except Exception as e:
        print("exception", i, e)
async def amain():
    loop = asyncio.get_event_loop()
    print("starting lockers")
    lockers = []
    for i in range(5):
        lockers.append(loop.create_task(locker(i)))
        await asyncio.sleep_ms(5)
    print("canceling lockers")
    for i in range(2):
        lockers[i].cancel()
    print("done")
    while True:
        await asyncio.sleep_ms(100)
asyncio.run(amain())
```
transcript (linux or esp32 both show it):
```
> ../../micropython/ports/unix/micropython test_lock.py
starting lockers
acquiring 0
got 0
acquiring 1
acquiring 2
acquiring 3
acquiring 4
canceling lockers
done
got 2
got 3
releasing 2
releasing 3
exception 3
got 4
releasing 4
```
Notice "got 3" between "got 2" and "releasing 2".

Looks like Kevin also posted his test case...

Reply:
Hmm, I need to go back and read the semantics of canceling, async with, etc. I'm probably misunderstanding what should happen but I find it strange that the transcript doesn't have any `exception [01]` in but an `exception 3`?

Reply:
Not sure where you expect "exception [01]" or what it is but exception 3 is when the 3rd locker releases the lock because after releasing 2 the lock is in a non-locked state and therefore the next call to lock.release() throws a runtime error

Reply:
Apologies for this getting long, but... here's an updated test case that runs in cpython, and micropython linux as well as esp32 (just need to adapt the path.append):
```
import sys
if sys.implementation.name == 'micropython':
    if sys.platform == 'linux': sys.path.append('/home/src/esp32/micropython/extmod')
    import uasyncio as asyncio
    loop = asyncio.get_event_loop()
    lock = asyncio.Lock()
else:
    import asyncio
    loop = asyncio.new_event_loop()
    lock = asyncio.Lock(loop=loop)
async def locker(i):
    try:
        print("acquiring", i)
        async with lock:
            print("got", i)
            await asyncio.sleep(0.100)
            print("releasing", i)
    except Exception as e:
        print("exception", i, e)
async def amain():
    print("starting lockers")
    lockers = []
    for i in range(5):
        lockers.append(loop.create_task(locker(i)))
        await asyncio.sleep(0.005)
    print("canceling lockers")
    for i in range(2):
        lockers[i].cancel()
    print("done")
    while True:
        await asyncio.sleep(0.100)
loop.run_until_complete(amain())
```

Reply:
To me it looks like task cancellation has a problem. I added a print statement into the `Lock`'s `except core.CancelledError` clause and it doesn't fire for locker 0, which is sitting in a sleep:
```
> ../../micropython/ports/unix/micropython test_lock.py
starting lockers
acquiring 0 <Task object at 7f83790cef40>
got 0
acquiring 1 <Task object at 7f83790cfce0>
acquiring 2 <Task object at 7f83790d02e0>
acquiring 3 <Task object at 7f83790d0320>
acquiring 4 <Task object at 7f83790d0460>
canceling lockers
done
task cancelled <Task object at 7f83790cfce0>
got 2
got 3
releasing 2
releasing 3
exception 3
got 4
releasing 4
```

Reply:
> Not sure where you expect "exception [01]" or what it is

I thought lockers 0 and 1 should hit the `except Exception as e`, I didn't realize that `asyncio.CancelledError` doesn't inherit from `Exception`.

Reply:
OK, took me a while to understand what's happening:
- task 0 gets the lock
- the others are blocked in the yield in acquire
- when task 0 gets canceled the __aexit__ of the async with lock clause happens to make task 2 runnable
- then task 1 gets canceled causing the except clause in acquire to trigger, it sees state==1 (which really is for task 2)  and thinks it's pending in resume and calls release which makes task 3 runnable
- now two tasks are runnable thinking they have the lock

Reply:
I don't see a way to fix this other than to add some form of "next_task" state variable and then something like `if self.state == 1 and self.next_task == core.cur_task:`

Reply:
I've now fixed this bug with `Lock` and added a test for it.  The fix is along the lines of what @kevinkk525 suggested, and what @tve suggested just above, but doesn't need both `state` and `next_task`/`_awt` variables, it just uses one state variable.